src/rulesets/LinearAlgebra/matfun.jl

```diff
@@ -129,7 +129,10 @@ function rrule(::typeof(exp), A0::StridedMatrix{<:BlasFloat})
         A = copy(A0)
         X, intermediates = _matfun!(exp, A)
         function exp_pullback(ΔX)
-            ∂A = _matfun_frechet_adjoint!(exp, ΔX, A, X, intermediates)
+            # Ensures ∂X is mutable. The outer `adjoint` is unwrapped without copy by
+            # the default _matfun_frechet_adjoint!
+            ∂X = Matrix(ΔX')'
```

Comment:
Weirdly `copy` promises to return something mutable and is probably cleaner than this?

Reply:
I guess `copy` would always allocate, where a type conversion won't? or should there be some dispatch using a mutable trait ?

Reply:
On checking, I was apparently wrong about `copy`.
It doesn't nesc return something mutable.
I was confused by the fact that it makes some view-like things (including `Adjoint` and `SubArray`, but not including `Diagonal`) into an arrays.

`convert` vs   Constructors is a thing though.

https://docs.julialang.org/en/v1/manual/conversion-and-promotion/#Mutable-collections

> `convert(T, x)` is expected to return the original `x` if `x` is already of type `T`. In contrast, if `T` is a mutable collection type then `T(x)` should always make a new collection (copying elements from `x`).

So perhaps this should be a `convert` if we want to avoid allocating unnesc?

---

ChainRulesCore does actually have a trait that might be suitable for this. If we really wanted.
It's part of the mechanics for doing inplace gradient accumulation


>   `is_inplaceable_destination(x) -> Bool`
> 
>   Returns true if `x` is suitable for for storing inplace accumulation of gradients.
>   For arrays this boils down `x .= y` if will work to mutate `x`, if `y` is an appropriate differential.
>   Wrapper array types do not need to overload this if they overload `Base.parent`, and are `is_inplaceable_destination` if and only if their parent array is.
>   Other types should overload this, as it defaults to `false`.

Reply:
I ended up going with a mixture of these both: Using the trait to decide if we should do anything and then using `convert`, which in this case will always allocate a copy. (`convert(Matrix, X')` will allocate unless `X` is an `Adjoint{T,Matrix{T}}`, but since that type is inplaceable, the trait will bypass the `convert` anyways).